# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
#in place function, returns but does not store
(lambda num : num + 2)(4)



6

#### Saving to a Variable

In [2]:
addTwoLambda = lambda num: num +2

addTwoLambda(10)

12

#### Multiple Inputs

In [3]:
# one-time in place
(lambda num1, num2, num3: num1 + num2 + num3)(6,4,10)

20

In [4]:
# creating a variable
add_three_numbers = lambda num1, num2, num3: num1 + num2 + num3

add_three_numbers(6,4,10)

20

In [ ]:
# function equivalent
def add_three_func(num1, num2, num3):
    return num1 + num2 + num3

#### Passing a Lambda into a Function

In [10]:
def outter_function():
    outter_num = 4
    def multiply(num):
        inner_num = 10
        return num * outter_num
    return multiply
print(outter_function()(4))

#assigning nested func to variable

my_multiply = outter_function()
my_multiply(4)

def return_Lambda(num1, num2):
    return lambda num3, num4 : num1 + num2 + num3 + num4

add_two_three = return_Lambda(2,3)

#return_Lambda(2,3)(10,20)
add_two_three(10, 20)

16


35

#### Returning a Lambda from a Function

In [14]:
def math_operations(f, num):
    return f(num)
math_operations(lambda x : x **2, 10)

math_operations(lambda num: num//10, 1000)

multiply_by_ten = lambda num : num * 10
math_operations(multiply_by_ten, 5)

50

In [21]:
def how_to_multiply(num1):
    return lambda num2 : num1 * num2

double = how_to_multiply(2)

triple = how_to_multiply(3)

double(10)

triple(10)

30

#### If Statements within Lambdas

In [17]:
# if requires an else

greater_than_10_add_2 = lambda num : num + 2 if num > 10 else num

# workaround with elif
greater_than_10_add_2 = lambda num : num + 2 if num > 10 else (num - 2 if num == 10 else num)

greater_than_10_add_2(11)

13

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [19]:
# cube if even, square if odd

f = lambda num : num**2 if num%2 == 1 else num**3
print(f(2))
print(f(3))

8
9


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
# map(<function>, iterable)

def square_number(num):
    return num**2

number_list = [1,2,3,4,5,6]

# list function
list(map(square_number, number_list))

[1, 4, 9, 16, 25, 36]

#### Using Lambda's with Map

In [29]:
# for loop with print statement
for num in (map(lambda num: num**3, number_list)):
    print(num, end=" ")

#creating a list inline
[num for num in map(lambda num: num**3, number_list)]

1 8 27 64 125 216 

[1, 8, 27, 64, 125, 216]

In [33]:
#able to pass multiple iterables to map, function must take multiple parameters

attackers = [5,6,2,8]
defenders = [4,7,1,7]

survivors = list(map(lambda attacker, defender: attacker < defender, attackers, defenders))
survivors.count(True) > survivors.count(False)

# .count(), loop through list and count occurences of argument

False

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [35]:
num_list = [1,2,3,4,5]
print(list(map(lambda num: num*2 -1, num_list)))



[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [38]:
# filter(<function, iterable) returns a filter object

def greater_than_9(num):
    return num>= 10

some_numbers = [10, 1, 100, 5, 11]

list(filter(greater_than_9, some_numbers))

[10, 100, 11]

#### Using Lambda's with Filter()

In [43]:
students = ['Hamed', 'Andy', 'William', 'Christopher', 'Will']

[student for student in filter(lambda student: student[0].lower() == 'w', students)]

['William', 'Will']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [49]:
from statistics import mean
#John's solution
listo_nums = [1,2,3,4,5,6]
print([num for num in filter(lambda num: num > mean(listo_nums), listo_nums)])

#YY's reverse solution
avg = mean(listo_nums)
below_avg = list(filter(lambda x: x < avg, listo_nums))
print(below_avg)


[4, 5, 6]
[1, 2, 3]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [51]:
# must import, not recommended
from functools import reduce

#reduce(<function>, iterable)

sum([1,2,3])

def add_2_nums(num1, num2):
    return num1+num2

reduce(add_2_nums, [1,2,3,4,5])




15

#### Using Lambda's with Reduce()

In [52]:
reduce(lambda num1, num2: num1*num2, [1,2,3,4,5])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [56]:
print(reduce(lambda num1, num2: num1-num2, [1,2,3,4,5]))

reduce(lambda num1, num2: num1**num2, [2,3,4,5])

-13


1152921504606846976

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [67]:
def add_num(num):
    if num == 1:
        print(f'add_num({num}) = 1')
        return 1
    print(f'add_num({num}) = ({num}) + add_num({num-1})')
    return num + add_num(num-1)
add_num(5)

add_num(5) = (5) + add_num(4)
add_num(4) = (4) + add_num(3)
add_num(3) = (3) + add_num(2)
add_num(2) = (2) + add_num(1)
add_num(1) = 1


15

#### Writing a Factorial Function

In [63]:
def find_factorial(num): 
    if num == 1:
        print(f'find_factorial({num}) = 1')
        return num
    print(f'find_factorial({num}) = ({num}) * find_factorial({num-1})')
    return num * find_factorial(num-1)
print(find_factorial(5))

find_factorial(5) = (5) * find_factorial(4)
find_factorial(4) = (4) * find_factorial(3)
find_factorial(3) = (3) * find_factorial(2)
find_factorial(2) = (2) * find_factorial(1)
find_factorial(1) = 1
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [1]:
def big_nums(num):
    if num >= 4:
        print('hey')
        return 4
    return num ** big_nums(num + 1)

big_nums(2)
    

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [5]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

import regex


print([place for place in filter(lambda place: len(place) > 2, places)])

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [21]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


sort_surnames = lambda namelist: namelist.sort(key = lambda name: name.split()[-1].lower())
    
sort_surnames(author)

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [91]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# for temp in places[len(places)][1]:
#     temp*(9/5)+32
# print(places)
# [list(place) for place in map(lambda place: place[1]*(9/5)+32, places)]

# places[place][1]

# places_f = [map(lambda place: places_f.append((place[0], place[1]*(9/5)+32)), places)]
# for place in map(lambda place:)
[place for place in map(lambda place: (place[0], place[1]*(9/5)+32), places)]

# print(places)



[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [104]:
def print_fib(num):
    for i in range(num+1):
        print('Iteration', str(i)+':', fib(i))

def fib(num):
    if num <= 1:
        return 1
    else:
        return fib(num -1)  + fib(num-2)
    
print_fib(5)


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
